File: core/python/src/aiddl_core/tools/stopwatch.py
import time


class StopWatch:
    m = {}
    start_time = {}

    def start(key):
        if key not in StopWatch.m.keys():
            StopWatch.m[key] = []
        StopWatch.start_time[key] = time.time()

    def stop(key):
        StopWatch.m[key] = time.time() - StopWatch.start_time[key]

    def get_sums_str():
        key_list = []
        for k in StopWatch.m.keys():
            key_list.append(k)
        key_list.sort()
        s = ""
        for k in key_list:
            s += "%s: %f" % (k, sum(StopWatch.m[k]))
        return s

File: core/scala/src/main/scala/org/aiddl/core/scala/util/StopWatch.scala
package org.aiddl.core.scala.util

import scala.collection.mutable.HashMap
import scala.collection.mutable

/**
 * Used to measure time elapsed between a start and a stop point.
 */
object StopWatch {
    private val startTimes = new HashMap[String, Long]

    /**
     * Recorded times
     */
    val recordedTimes = new HashMap[String, List[Long]]

    /**
     * Start the stop watch for a name
     * @param name name of the interval
     */
    def start( name: String ): Unit = startTimes.put(name, System.nanoTime)

    /**
     * Stop the stop watch for a name and record elapsed time in list of all recorded times for the given name.
     * @param name name of the interval
     * @return time in seconds since start was called
     */
    def stop(  name: String ): Double = {
        val list = recordedTimes.getOrElseUpdate(name, Nil)
        val duration = (System.nanoTime - startTimes(name))
        recordedTimes.put(name, duration :: list )
        duration / 1000000000.0
    }

    /**
     * Create a summary containing the sum and number of records for every recorded name
     * @return summary as a string
     */
    def summary: String = {
        val sb = new StringBuilder
        var keys = recordedTimes.keys.toSeq
        keys.sorted.foreach( k => { 
            val sum = recordedTimes(k).foldLeft(0.0)((c, t) => c + t/1000000000.0)
            sb.addAll(s"$k: $sum s (${recordedTimes(k).size} entries)\n")})

        sb.toString
    }

}
